File: bwu/dp/264-ugly-number-ii.py
class Solution(object):
    def nthUglyNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        uglys = [1]
        indices = [0] * 3
        primes = [2, 3, 5]
        
        while len(uglys) < n:
            next = min(primes[i] * uglys[indices[i]] for i in xrange(3))
            if next / uglys[indices[0]] == 2:
                indices[0] += 1
            if next / uglys[indices[1]] == 3:
                indices[1] += 1
            if next / uglys[indices[2]] == 5:
                indices[2] += 1
                
            if uglys[-1] == next:
                continue
            
            uglys.append(next)
            
        return uglys[n - 1]

File: captainhcg/go/264-ugly-number-ii.go
func nthUglyNumber(n int) int {
    arr := []int{1}
    p2, p3, p5 := 0, 0, 0
    for n > 1{
        un := min(arr[p2] * 2, arr[p3] * 3, arr[p5] * 5)
        arr = append(arr, un)
        if un % 2 == 0{
            p2 += 1
        }
        if un % 3 == 0{
            p3 += 1
        }
        if un % 5 == 0{
            p5 += 1
        }
        n -= 1
    }
    return arr[len(arr)-1]
}

func min(v1 int, vn ...int) (m int) {
    m = v1
    for _, v := range vn{
        if v < m{
            m = v
        }
    }
    return
}
